libraries/SPI/SPI.cpp

```diff
@@ -235,155 +235,230 @@ void SPIClass::transfer(void *buf, size_t count)
+{
+    uint8_t *txbuf8 = (uint8_t *)txbuf,
+            *rxbuf8 = (uint8_t *)rxbuf;
+    if(rxbuf8) {
+        if(txbuf8) {
+            // Writing and reading simultaneously
+            while(count--) {
+                *rxbuf8++ = _p_sercom->transferDataSPI(*txbuf8++);
+            }
+        } else {
+            // Reading only
+            while(count--) {
+                *rxbuf8++ = _p_sercom->transferDataSPI(0xFF);
+            }
+        }
+    } else if(txbuf) {
+        // Writing only
+        while(count--) {
+            (void)_p_sercom->transferDataSPI(*txbuf8++);
+        }
+    }
+}
+
 // Pointer to SPIClass object, one per DMA channel.
 static SPIClass *spiPtr[DMAC_CH_NUM] = { 0 }; // Legit inits list to NULL
 
-void SPIClass::dmaCallback(Adafruit_ZeroDMA *dma) {
-  // dmaCallback() receives an Adafruit_ZeroDMA object. From this we can get
-  // a channel number (0 to DMAC_CH_NUM-1, always unique per ZeroDMA object),
-  // then locate the originating SPIClass object using array lookup, setting
-  // the dma_busy element 'false' to indicate end of transfer.
-  spiPtr[dma->getChannel()]->dma_busy = false;
+// dma callback when the read part is completed
+void SPIClass::dmaCallback_read(Adafruit_ZeroDMA *dma)
+{
+	// dmaCallback() receives an Adafruit_ZeroDMA object. From this we can get
+	// a channel number (0 to DMAC_CH_NUM-1, always unique per ZeroDMA object),
+	// then locate the originating SPIClass object using array lookup, setting
+	uint8_t channel = dma->getChannel();
+
+	// flag this part of the dma done
+	spiPtr[channel]->dma_read_done = true;
+
+	// see if the entire transaction is completed
+	checkDmaComplete(channel);
 }
 
-void SPIClass::transfer(const void* txbuf, void* rxbuf, size_t count,
-  bool block) {
+// dma callback when the write part is completed
+void SPIClass::dmaCallback_write(Adafruit_ZeroDMA *dma)
+{
+	// dmaCallback() receives an Adafruit_ZeroDMA object. From this we can get
+	// a channel number (0 to DMAC_CH_NUM-1, always unique per ZeroDMA object),
+	// then locate the originating SPIClass object using array lookup, setting
+	uint8_t channel = dma->getChannel();
+
+	// flag this part of the dma done
+	spiPtr[channel]->dma_write_done = true;
+
+	// see if the entire transaction is completed
+	checkDmaComplete(channel);
+
+}
+
+// see if the entire dma transaction is completed
+// will automatically initiate another dma if we have bytes remaining to transfer
+void SPIClass::checkDmaComplete(uint8_t channel)
+{
+	// read and write dmas are both done
+	if(spiPtr[channel]->dma_read_done && spiPtr[channel]->dma_write_done)
+	{
+		// are more bytes that need to be transfered
+		// fire another dma transaction
+		if( (spiPtr[channel]->dma_bytes_remaining) > 0)
+		{
+			// initiate another transfer for the next section of bytes
+			// update buffer pointers offsets
+			// use the same user callback as last time if any
+			void* txbuf = spiPtr[channel]->txbuf_last + DMA_MAX_TRANSFER_SIZE;
```

Comment:
warning: pointer of type 'void *' used in arithmetic [-Wpointer-arith]
    void* txbuf = spiPtr[channel]->txbuf_last + DMA_MAX_TRANSFER_SIZE;

This and the next line generate warnings. I think casting to explicit pointer types should solve it

Reply:
Ok added in the type cast :-)